src/main/java/io/quarkus/gizmo2/GenericType.java

```diff
@@ -669,8 +669,25 @@ public ClassDesc desc() {
         }
 
         public StringBuilder toString(final StringBuilder b) {
-            // supertype annotations []
-            return super.toString(componentType.toString(b)).append("[]");
+            GenericType element = this;
+            while (element instanceof OfArray arr) {
+                element = arr.componentType;
+            }
+            element.toString(b);
+            appendComponent(b);
+            return b;
+        }
+
+        private void appendComponent(StringBuilder b) {
+            // cannot use just `hasAnnotations()`, because that delegates to `has{Visible,Invisible}Annotations()`,
+            // which are both overridden in this class to also check the component type
+            if (super.hasVisibleAnnotations() || super.hasInvisibleAnnotations()) {
+                b.append(' ');
+            }
```

Comment:
This _possibly maybe_ suggests that the overrides of `hasVisibleAnnotations()` and `hasInvisibleAnnotations()` in this class are not correct. They look like:

```java
public boolean hasVisibleAnnotations() {
    return super.hasVisibleAnnotations() || componentType.hasVisibleAnnotations();
}

public boolean hasInvisibleAnnotations() {
    return super.hasInvisibleAnnotations() || componentType.hasInvisibleAnnotations();
}
```

I have no idea why they are like this -- it is my opinion that these methods shouldn't be overridden in this class at all, but I also guess there's a reason behind it.

Reply:
Sure.